### PR TITLE
Filter by time based availability

### PR DESCRIPTION
**Pull request by kpuputti**
This PR adds time-based availability filtering to the listing search.

There is a new `BookingDateRangeLengthFilter` filter component where the user can select start/end dates, and an optional min duration for the booking. The min duration can only be selected if the dates are selected.

The new filter can be controlled in the `marketplace-custom-config.js`:

```js
// Activate booking dates filter on search page
export const dateRangeLengthFilterConfig = {
  active: true,

  // A global time zone to use in availability searches. As listings
  // can be in various time zones, we must decide what time zone we
  // use in search when looking for available listings within a
  // certain time interval.
  //
  // If you have all/most listings in a certain time zone, change this
  // config value to that.
  //
  // See: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
  searchTimeZoneId: 'Etc/UTC',

  // Options for the minimum duration of the booking
  options: [
    { key: '0', label: 'Any length' },
    { key: '60', label: '1 hour', shortLabel: '1h' },
    { key: '120', label: '2 hours', shortLabel: '2h' },
  ],
};
```

## Mobile filter

Screenshot 2020-04-15 at 14 57 30

## Desktop filter

Screenshot 2020-04-15 at 14 57 15

### CODE REVIEW

File: src/containers/SearchPage/MainPanel.js [hunk 1]
```diff
@@ -64,8 +64,10 @@ class MainPanel extends Component {
       : {};
 
     const hasPaginationInfo = !!pagination && pagination.totalItems != null;
-    const totalItems = searchParamsAreInSync && hasPaginationInfo ? pagination.totalItems : 0;
-    const listingsAreLoaded = !searchInProgress && searchParamsAreInSync && hasPaginationInfo;
+    const listingsLength = listings ? listings.length : 0;
+    const totalItems =
+      searchParamsAreInSync && hasPaginationInfo ? pagination.totalItems : listingsLength;
```
Comment (Gnito) [inline src/containers/SearchPage/MainPanel.js:69]: So, the decision was that `totalItems` is not returned in this case?
If that's true, you should write some comment about this exceptional API behaviour.

File: src/containers/SearchPage/SearchPage.duck.js [hunk 2]
```diff
@@ -131,15 +132,42 @@ export const searchListings = searchParams => (dispatch, getState, sdk) => {
...
+    const startDate = hasDateValues ? dateValues[0] : null;
+    const endDate = hasDateValues ? dateValues[1] : null;
+    const minDurationMaybe =
+      minDurationParam && Number.isInteger(minDurationParam) && hasDateValues
+        ? { minDuration: minDurationParam }
+        : {};
+
+    return hasDateValues
+      ? {
+          start: formatDateStringToUTC(startDate),
```
Comment (Gnito) [inline src/containers/SearchPage/SearchPage.duck.js:147]: This doesn't work. UTC is not used in hourly at all. Or is the API now unexpectedly expecting UTC in here? 

More questions: what kind of strings are used in the URL params? Do they contain time zone?

Anyway, FTW-hourly uses listing's time zone always, when presenting dates - and Flex transaction process doesn't care about time zones with time-based process just timestamps (except in availabilityPlan where listing's time zone is saved so that availability of "Tue 09:00 - 17:00" is understood correctly.)

So, If the listing is in NY and customer is in Helsinki, you need to adjust the timestamp so that start (provided by customer's browser) is converted to the same time-of-day in NY's time zone (which is a different timestamp).

So, the question is now that should that be done before the time is visible in the URL params or here. My gut feeling is that these URL's will be shared and therefore URL params should already reflect the listing's time zone or the time zone info should be completely absent.

File: src/forms/FilterForm/FilterForm.js [hunk 3]
```diff
@@ -58,6 +58,25 @@ const FilterFormComponent = props => {
 
         const classes = classNames(css.root);
 
+        const spy =
+          liveEdit || onChange ? (
+            <FormSpy onChange={handleChange} subscription={{ values: true, dirty: true }} />
```
Comment (Gnito) [inline src/forms/FilterForm/FilterForm.js:63]: My experience with these FormSpy setups is that they are extremely tricky. Check couple of times (even though this is an existing setup).

File: src/containers/SearchPage/MainPanel.js (hunk 1 above)
Comment (kpuputti) [reply src/containers/SearchPage/MainPanel.js:69]: Here's an example meta object in the response:

Screenshot 2020-04-09 at 16 27 43

File: src/containers/SearchPage/SearchPage.duck.js (hunk 2 above)
Comment (kpuputti) [reply src/containers/SearchPage/SearchPage.duck.js:147]: I don't quite understand how the listing time zone plays here. The user is making a query for listings, so there can be various time zones in the result.

What time zone should the query be in then? Not sure. Are our options either UTC or the user's time zone? Or is there a configured time zone that is "global" to the marketplace?

File: src/containers/SearchPage/MainPanel.js (hunk 1 above)
Comment (Gnito) [reply src/containers/SearchPage/MainPanel.js:69]: Use that `paginationUnsupported` info in `const hasPaginationInfo`
and add some comment there about the reason.

File: src/containers/SearchPage/SearchPage.duck.js (hunk 2 above)
Comment (Gnito) [reply src/containers/SearchPage/SearchPage.duck.js:147]: @kpuputti 

If the Yoga session is held in Bali (i.e. customer tries to find yoga class for her vacation), then her assumption is that "when I'm on Bali I want to have yoga class at 10 am."

However, since that aforementioned customer is not at the Bali at the time of reservation but somewhere else, the Date object she gets from a browser is not pointing to correct time-of-day in Bali.

Let's say the customer is from Helsinki.
```
> local10am = new Date('2020-04-09 10:00:00')
// Thu Apr 09 2020 10:00:00 GMT+0300 (Eastern European Summer Time)
```

Use util/dates.js (`timeOfDayFromLocalToTimeZone`) to convert the time-of-day between time zones:
```shell
> listingTZ10am = timeOfDayFromLocalToTimeZone(local10am, 'Asia/Makassar')
// Browser would print that timestamp as:
// Thu Apr 09 2020 05:00:00 GMT+0300 (Eastern European Summer Time)
// because the 10 am in Bali is the same moment as 05 am in Helsinki.
```

If you want to render the timestamp correctly in listing's time zone in UI, there's util/dates helper function for that too, but for the sake of this example, I'll use Web Intl API directly:
```
> options = {
    year: 'numeric', month: 'numeric', day: 'numeric',
    hour: 'numeric', minute: 'numeric', second: 'numeric', 
    timeZone: 'Asia/Makassar',
    timeZoneName: 'short'
  };
> console.log(new Intl.DateTimeFormat('fi-FI', options).format(listingTZ10am));
// 9.4.2020 klo 10.00.00 UTC+8
```

File: src/translations/en.json [hunk 4]
```diff
@@ -75,6 +75,11 @@
   "BookingDateRangeFilter.labelPopup": "Dates",
   "BookingDateRangeFilter.labelSelectedPlain": "{dates}",
   "BookingDateRangeFilter.labelSelectedPopup": "{dates}",
+  "BookingDateRangeLengthFilter.labelPlain": "Dates",
+  "BookingDateRangeLengthFilter.labelPopup": "Dates",
+  "BookingDateRangeLengthFilter.labelSelectedPlain": "{dates}",
+  "BookingDateRangeLengthFilter.labelSelectedPopup": "{dates}",
+  "BookingDateRangeLengthFilter.minDurationLabel": "Minimum length",
```
Comment (kpuputti) [inline src/translations/en.json:82]: This is a new translation key. Should this line be added to the other translations as well?

File: src/containers/SearchPage/SearchPage.js [hunk 5]
```diff
@@ -222,10 +228,11 @@ export class SearchPageComponent extends Component {
             searchParamsForPagination={parse(location.search)}
             showAsModalMaxWidth={MODAL_BREAKPOINT}
             primaryFilters={{
-              yogaStylesFilter: filters.yogaStylesFilter,
-              certificateFilter: filters.certificateFilter,
               priceFilter: filters.priceFilter,
+              dateRangeLengthFilter: filters.dateRangeLengthFilter,
               keywordFilter: filters.keywordFilter,
+              yogaStylesFilter: filters.yogaStylesFilter,
+              certificateFilter: filters.certificateFilter,
```
Comment (kpuputti) [inline src/containers/SearchPage/SearchPage.js:235]: The Yoga styles and Teacher certificate filters will be moved to secondary filters in another PR.

File: src/forms/FilterForm/FilterForm.js (hunk 3 above)
Comment (kpuputti) [reply src/forms/FilterForm/FilterForm.js:63]: Not sure what I should've looked for, but haven't noticed any problems this far.

File: src/translations/en.json (hunk 4 above)
Comment (kpuputti) [reply src/translations/en.json:83]: @Gnito See the questions above ^

File: src/containers/SearchPage/SearchPage.js (hunk 5 above)
Comment (kpuputti) [reply src/containers/SearchPage/SearchPage.js:235]: @Gnito ^

**Comment by Gnito** [comment]
"Could not find any teachers." text goes behind the map in 768px.

File: src/translations/en.json (hunk 4 above)
Comment (Gnito) [reply src/translations/en.json:83]: I'm not sure what you mean. Options are currently in marketplace-custom-config.js to make it easier to change them on the fly (especially, with category and amenities it is useful). However, since this is not about those, it's probably best if this title-label is here in the translation file.

Comment (Gnito) [reply src/translations/en.json:83]: Or did you mean that es, de, fr, should be changed too?

That's happening through WTI at some point (when someone has time to take care of FTW and ask new translations from translators).

Comment (kpuputti) [reply src/translations/en.json:83]: Yeah, that's what I meant. This is a new translation key, but there's only the English text.

File: src/util/dates.js [hunk 6]
```diff
@@ -623,6 +623,26 @@ export const formatDateStringToUTC = dateString => {
...
+ * ('2020-04-15', 'Europe/Helsinki') => '2020-04-14T21:00:00.000Z'
+ *
+ * @param {String} dateString in 'YYYY-MM-DD' format
+ * @param {String} tzId time zone id, see:
+ *   https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+ *
+ * @returns {String} string in '0000-00-00T00:00:00.000Z' format
+ */
+export const formatDateStringToTz = (dateString, tzId) => {
+  return moment.tz(dateString, tzId).toDate();
```
Comment (Gnito) [inline src/util/dates.js:643]: Other functions in this util file use 'timeZone' instead of 'tzId'. It's probably better to be consistent.

File: src/util/dates.js [hunk 7]
```diff
@@ -641,6 +661,25 @@ export const getExclusiveEndDate = dateString => {
...
+ * the given time zone. Adds 1 day to work with the exlusive date
+ * range in the API.
+ *
+ * @param {String} dateString in 'YYYY-MM-DD'format
+ * @param {String} tzId time zone id, see:
+ *   https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+ *
+ * @returns {String} string in '0000-00-00T00:00:00.000Z' format
+ */
+export const getExclusiveEndDateWithTz = (dateString, tzId) => {
```
Comment (Gnito) [inline src/util/dates.js:675]: I'd change tzId here too. (because of consistency)

File: src/translations/en.json (hunk 4 above)
Comment (Gnito) [reply src/translations/en.json:83]: FTW defaults to en.json if the key is completely missing from other translation files - so, if the correct translations are not quick to collect, you can leave those as they are.

Comment (Gnito) [reply src/translations/en.json:83]: I.e. don't add English translations to fr/de/es.json files.

File: src/marketplace-custom-config.js [hunk 8]
```diff
@@ -26,8 +26,26 @@ export const priceFilterConfig = {
...
+  // A global time zone to use in availability searches. As listings
+  // can be in various time zones, we must decide what time zone we
+  // use in search when looking for available listings within a
+  // certain time interval.
+  //
+  // If you have all/most listings in a certain time zone, change this
+  // config value to that.
+  //
+  // See: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+  searchTimeZoneId: 'Etc/UTC',
```
Comment (Gnito) [inline src/marketplace-custom-config.js:41]: We should probably mention something about this in 
https://www.sharetribe.com/docs/cookbook-search/change-search-filters-in-ftw/#filter-types

File: src/containers/SearchPage/SearchPage.duck.js [hunk 9]
```diff
@@ -131,12 +132,37 @@ export const searchListings = searchParams => (dispatch, getState, sdk) => {
...
+        ? { minDuration: minDurationParam }
+        : {};
+
+    const tzId = config.custom.dateRangeLengthFilterConfig.searchTimeZoneId;
+
+    return hasDateValues
+      ? {
+          start: formatDateStringToTz(startDate, tzId),
+          end: getExclusiveEndDateWithTz(endDate, tzId),
+          availability: 'time-partial',
```
Comment (Gnito) [inline src/containers/SearchPage/SearchPage.duck.js:152]: Maybe these hard-coded values could use some comment text or link to api-reference.

We have added this to FTW-daily: https://github.com/sharetribe/ftw-daily/blob/master/src/containers/SearchPage/SearchPage.duck.js#L147

File: src/containers/SearchPage/MainPanel.js [hunk 10]
```diff
@@ -63,9 +63,18 @@ class MainPanel extends Component {
...
 
-    const hasPaginationInfo = !!pagination && pagination.totalItems != null;
-    const totalItems = searchParamsAreInSync && hasPaginationInfo ? pagination.totalItems : 0;
-    const listingsAreLoaded = !searchInProgress && searchParamsAreInSync && hasPaginationInfo;
+    // With time-based availability filtering, pagination is NOT
+    // supported. In these cases we get the pagination support info in
+    // the response meta object, and we can use the count of listings
+    // as the result count.
+    //
+    // See: https://www.sharetribe.com/api-reference/marketplace.html#availability-filtering
```
Comment (Gnito) [inline src/containers/SearchPage/MainPanel.js:71]: 👍

File: src/marketplace-custom-config.js (hunk 8 above)
Comment (kpuputti) [reply src/marketplace-custom-config.js]: Sure, I have another story for doc updates. I added that link to the story.
